### ISSUE TITLE
Refining the Route for Intent Driven Apps

### ISSUE THREAD

**Issue by HighOnDrive** [TylorS/cycle-router#1]
The following insightful article resonates with me. In it Christian Alfoni dissected the idea and purpose of a router and a route. He ended by saying that we need "one awesome concept for all the state changes of our applications".

http://www.christianalfoni.com/articles/2015_08_20_What-if-the-adddressbar-worked-like-an-input

There was also the following project that provided a broader definition of routing. It introduced itself by saying it was "built on learnings from large applications that ties together routes, states, business logic, and React components".

https://github.com/daftdevelopers/react-scaffold#architecture

What I love about Cycle,js is that it is intent driven and coupled with RxJS it is a perfect base for modern Human-Computer Interaction apps. However, there is still the need to manage all the interaction and it is difficult to do that with routers that define a route as just being a location. 

In the past simple clicks on links were the extent of interaction so the way routers always worked was good enough. Now we have many event sources and RxJS processing in the picture plus nested components and routing to be useful needs a revamp. Intent driven apps need a more robust router and more elaborate routes. 

My idea for this involves a JSON configuration document I call a Manifest, which is also a representation of the entire app state. The Manifest contains sections for each of the major tool modes that an app can have. Because interaction is the primary mover in an intent driven app components are then just optional parameters of an interaction task. The various tools of the app are then the ultimate route definers, defining what aspects they need to have activated to carry out their intended function.

I see the Manifest idea echoed in the following issue too, especially this statement "This is an interesting idea... I think that having some way to store the stack outside of the Navigator would extremely useful -- I'd like to be able to store it within a Redux store, for example, so I can save my entire app state in Redux and reload from there."

https://github.com/facebook/react-native/issues/2436

Here is how the Manifest based routing would then play out in the most generic way possible:

1) The app would load a Manifest of sections that define what an app can do or does. The Manifest defines a flexible exchangeable version of the app. Presently, the Cycle-Router example defines routes and handlers in a hardwired way in code. Instead, each section would declare its path and handler and other optional parameters as key:values in a softwired configurable way.
 
2) Each section can have an optional value that defines its trigger point/s in the layout of a given top level or nested component.

3) A simple hit detection function then works with any custom UI layout to test for trigger points (in desktop Electron apps qwerty key triggers can also be mapped or gestures can be used from some other controller).

4) If events originate in a menu zone or are the right trigger via some gesture then lookup the corresponding section in the active Manifest.

5) Activate the section and handle each of its optional parameters. Parameters include: what tool (state modifier) subscribes to the event stream coming off the components intent function, what nested component should be injected into the base surface of the app, what stylesheet (CSS module) to use and inject into the optional component and any other state that is suitable for defining what the interaction flow and look of the app should be in a specific tool route. Parameters in a section can also be set so sections also work as state trackers for the context of a given tool interaction mode.

That is an overview of how I would like routes and the route processor to work in a super flexible way. There are many more details, which I can provide if Cycle-Router might like to implement the Manifest approach to interaction management.

For this to work smoothly the following scope issue would likely need to be resolved, because sections can all behave uniquely, some may trigger more long lived editor components while others trigger more transient components.

https://github.com/cyclejs/cycle-core/issues/167

P.S. If there is interest in building a Manifest configurable router then a next step might be to develop a app template for larger Cycle.js apps that puts Cycle-Router into practice.

**Comment by TylorS** [comment]
I definitely don't think I'm opposed to going in this direction. I would certainly like for this router to be more than just a *traditional* router. I also want to make sure that it continues to and/or becomes more 'idiomatic' in relation to the way cycle apps **should** be written, which I'm certainly no expert in yet.  

I think it would be helpful to see what one of these Manifests to get a better feel for how you would propose they would look. Also maybe some sort of mock of how this would tie into a cycle application with and/or without cyclejs/cycle-core#167 being resolved.

**Comment by HighOnDrive** [comment]
Presently there are no examples of a larger Cycle app. To move in that direction I see the following:

1) Mechanics = the Cycle API
2) Context = an app structure using MVI (the mock)
3) Strategy = interaction state manager (the manifest)

The strategy introduced above works within an app context, so I agree it helps to know what that is! I'll get back to you on that...

**Comment by Frikki** [comment]
I like to think in terms of node-graphs. Eventually, I could envision a manifest being a node-tree. Perhaps, to stay in the paradigm of functional programming, each node could be a function that does whatever the designer pleases.

**Comment by HighOnDrive** [comment]
It is a node-tree and a graph. I'll work it, share it and do appreciate all the feedback I can get. Over all I'm interested in harnessing the powers of Cycle for some practical and generic structured app template. One that is flexible and can be customized and extended.

**Comment by bzalasky** [comment]
Client-side routing isn't that complicated, and I would consider it a solved problem for the purposes of most real-world applications (with today's browsers). Nothing has fundamentally changed around this because of React, Cycle or any other recent framework. All I want from my routing solution is to be able to specify a config object with paths as keys and a handlers as values, so that I can get on with building the actual features that will delight the users of my application.

Consider [Marionette's AppRouter](http://marionettejs.com/docs/v2.4.2/marionette.approuter.html#configure-routes) (simply as an example of something that has worked well for me in the past). It's simple and effective, and has unit tests. Within a Cycle application, I'd like to write a routes.js file that imports my handlers, and exports the routes configuration object. This could be passed to the router driver when the app is initialized. Exposing an observable to track when routes change is something that can be added, but isn't really an essential feature. I'd focus on solving the core problem, and develop a solution that's thoroughly tested as opposed to inventing something new.

Can you provide more details on the application you're building and why you need something beyond the scope of the router I'm describing above?

**Comment by bzalasky** [comment]
Being able to register multiple configuration objects could allow you to segment handlers and routes by component, if that's desirable for your application. Though you'd be surprised at how far you can get with a wildcard route and a whitelist. I'd approach this problem from the base use case and expand from there.

**Comment by HighOnDrive** [comment]
Great suggestions and bang on with our own thoughts on the matter! Can't see why it won't work just as you like. Thanks for having such insights and sharing them.

**Comment by TylorS** [comment]
Absolutely. I agree with you 100%. I guess you could call what I'm working
towards now is a minimal viable product. At first it will do just the
basics you ask. Declared like so, `'home': {path: '/home', handlers:
function()}` (open for api change and discussion). Handlers will also
except an array of functions. I want to make sure that this works first and
can be configurable within each subcomponent/nested dialogue, hopefully as
close as possible to the 'This component is the entire world' way of
things. Once that is 100% and good to go features could be added, possibly
as plugins.
